pydantic_ai_slim/pydantic_ai/mcp.py

```diff
@@ -255,11 +255,79 @@ async def direct_call_tool(
             # The MCP SDK wraps primitives and generic types like list in a `result` key, but we want to use the raw value returned by the tool function.
             # See https://github.com/modelcontextprotocol/python-sdk#structured-output
             if isinstance(structured, dict) and len(structured) == 1 and 'result' in structured:
-                return structured['result']
-            return structured
+                return (
+                    messages.ToolReturn(return_value=structured['result'], metadata=result.meta)
+                    if result.meta
+                    else structured['result']
+                )
+            return messages.ToolReturn(return_value=structured, metadata=result.meta) if result.meta else structured
 
-        mapped = [await self._map_tool_result_part(part) for part in result.content]
-        return mapped[0] if len(mapped) == 1 else mapped
+        mapped_part_metadata_tuple_list = [await self._map_tool_result_part(part) for part in result.content]
+        if (
+            all(mapped_part_metadata_tuple[1] is None for mapped_part_metadata_tuple in mapped_part_metadata_tuple_list)
```

Comment:
I think this can be `all(metadata is None for (part, metadata) in mapped_part_metadata_tuple_list)`. Let's use tuple destructuring everywhere instead of index-based reading.

Also let's move this to a `parts_have_metadata = any(...)` variable (inverted to be positive) as we use it multiple times (e.g. `if not parts_have_metadata and result.metadata is None` reads pretty clearly)

Reply:
@anirbanbasu Yep sounds good